### PR TITLE
Add search and cosine distance for brute-force

### PR DESCRIPTION
**Pull request by PetrovKP**
- Enable cosine metric for KNN in oneAPI interface
- Enable search task for KNN
- Fix performance gaps for brute-force
- Add example for search with cosine metric

TODO for next PR:
- Add tests
- Add docs

### CODE REVIEW

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i [hunk 1]
```diff
@@ -271,19 +277,29 @@ protected:
         {
             for (size_t kk = 0; kk < k; ++kk)
             {
-                // max(0, d) to remove negative distances before Sqrt
-                kDistances[i * k + kk] = services::internal::max<cpu, FPType>(FPType(0), heaps[i][kk].distance);
-                kIndexes[i * k + kk]   = heaps[i][kk].index;
+                if (distancesInstance->getType() == PairwiseDistanceType::euclidean)
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i:280]: Keep for minkowski as well?
I know that now there are abs() for them. But can be changed, probably, in the future.

Comment (SmirnovEgorRu) [reply cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i:280]: + Probably, for keeping the performance - better to move this if outside the loop

File: cpp/daal/src/algorithms/service_kernel_math.h [hunk 2]
```diff
@@ -270,6 +282,48 @@ class EuclideanDistances : public PairwiseDistances<FPType, cpu>
...
+        const FPType * const aa = super::normBufferA.get() + aOffset;
+        const FPType * const bb = (&(super::_a) == &(super::_b)) ? super::normBufferA.get() + bOffset : super::normBufferB.get() + bOffset;
+
+        PRAGMA_IVDEP
+        PRAGMA_VECTOR_ALWAYS
+        for (size_t i = 0; i < nRowsC; i++)
+        {
+            for (size_t j = 0; j < nColsC; j++)
+            {
+                res[i * nColsC + j] = FPType(1) - (res[i * nColsC + j] / Math<FPType, cpu>::sSqrt(aa[i] * bb[j]));
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/algorithms/service_kernel_math.h:320]: sSqrt can be to slow, better to try vSqrt somehow

Comment (SmirnovEgorRu) [reply cpp/daal/src/algorithms/service_kernel_math.h:320]: For example use TLS for tmp sqrt results

Comment (PetrovKP) [reply cpp/daal/src/algorithms/service_kernel_math.h:320]: @SmirnovEgorRu 
and division too. I want to move into init and division and root. but later. will be:
```suggestion
                res[i * nColsC + j] = FPType(1) - (res[i * nColsC + j] * aa[i] * bb[j]));
```

**Comment by PetrovKP** [comment]
http://intel-ci.intel.com/ebcc3a78-aff9-f172-826c-8cdcd4b723a1

**Comment by PetrovKP** [comment]
http://intel-ci.intel.com/ebcdd6b9-17c4-f1fa-9b95-8cdcd4b723a1

**Comment by PetrovKP** [comment]
http://intel-ci.intel.com/ebce14b0-9479-f112-8d81-8cdcd4b723a1

**Comment by PetrovKP** [comment]
http://intel-ci.intel.com/ebce29d6-f47c-f1fc-9475-8cdcd4b723a1

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i (hunk 1 above)
Comment (PetrovKP) [reply cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i]: done

File: cpp/daal/src/algorithms/service_kernel_math.h [hunk 3]
```diff
@@ -280,6 +307,88 @@ class EuclideanDistances : public PairwiseDistances<FPType, cpu>
...
+        super::computeABt(a, b, nRowsA, nColsA, nRowsB, res);
+
+        const FPType * const aa = super::normBufferA.get() + aOffset;
+        const FPType * const bb = (&(super::_a) == &(super::_b)) ? super::normBufferA.get() + bOffset : super::normBufferB.get() + bOffset;
+
+        PRAGMA_IVDEP
+        PRAGMA_VECTOR_ALWAYS
+        for (size_t i = 0; i < nRowsC; i++)
+        {
+            for (size_t j = 0; j < nColsC; j++)
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/algorithms/service_kernel_math.h:345]: Better to move pragmas to this loop

File: cpp/daal/src/algorithms/service_kernel_math.h [hunk 4]
```diff
@@ -280,6 +307,88 @@ class EuclideanDistances : public PairwiseDistances<FPType, cpu>
...
+                FPType sum = FPType(0);
+                PRAGMA_IVDEP
+                PRAGMA_ICC_NO16(omp simd reduction(+ : sum))
+                for (size_t j = 0; j < nCols; j++)
+                {
+                    sum += data[i * nCols + j] * data[i * nCols + j];
+                }
+                r[i] = sum;
+            }
+            Math<FPType, cpu>::vSqrt(blockSizeReal, r, r);
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/algorithms/service_kernel_math.h:385]: So, only one difference between computeNorm for Euclidean and Cosine - this Sqrt?
If so, better to extract common code somehow, even ```if``` is better

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i [hunk 5]
```diff
@@ -72,29 +75,32 @@ public:
...
+            newTrainLabelTable->getBlockOfRows(0, nTrain, readOnly, trainLabelBlock);
             trainLabel = trainLabelBlock.getBlockPtr();
             DAAL_CHECK_MALLOC(trainLabel);
         }
 
-        services::SharedPtr<daal::algorithms::internal::PairwiseDistances<FPType, cpu> > dist;
+        services::SharedPtr<PairwiseDistances<FPType, cpu> > dist;
 
-        switch (pairwiseDistance)
+        if (pairwiseDistance == PairwiseDistanceType::minkowski && minkowskiDegree == 2.0)
```
Comment (averbukh) [inline cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i:85]: Switch is preferred here (with 'if' in minkowski case)

File: cpp/daal/src/algorithms/service_kernel_math.h [hunk 6]
```diff
@@ -116,7 +126,9 @@ class EuclideanDistances : public PairwiseDistances<FPType, cpu>
 
     virtual ~EuclideanDistances() {}
 
-    virtual services::Status init()
+    PairwiseDistanceType getType() override { return PairwiseDistanceType::euclidean; }
```
Comment (averbukh) [inline cpp/daal/src/algorithms/service_kernel_math.h:129]: Use DAAL_C11_OVERRIDE instead (also for other functions)

File: cpp/daal/src/algorithms/service_kernel_math.h [hunk 7]
```diff
@@ -280,6 +307,88 @@ class EuclideanDistances : public PairwiseDistances<FPType, cpu>
     TArray<FPType, cpu> normBufferB;
 };
 
+// compute: A*B' / (sum(A^2, 2) * sum(B^2, 2))
+template <typename FPType, CpuType cpu>
+class CosineDistances : public EuclideanDistances<FPType, cpu>
```
Comment (averbukh) [inline cpp/daal/src/algorithms/service_kernel_math.h:312]: Why public inheritance from EuclideanDistances?

File: cpp/daal/src/algorithms/service_kernel_math.h [hunk 8]
```diff
@@ -280,6 +307,88 @@ class EuclideanDistances : public PairwiseDistances<FPType, cpu>
...
+{
+public:
+    using super = EuclideanDistances<FPType, cpu>;
```
Comment (averbukh) [inline cpp/daal/src/algorithms/service_kernel_math.h:315]: Makes sense to have it private/protected instead.

File: cpp/daal/src/algorithms/service_kernel_math.h [hunk 9]
```diff
@@ -280,6 +307,88 @@ class EuclideanDistances : public PairwiseDistances<FPType, cpu>
...
+
+    CosineDistances(const NumericTable & a, const NumericTable & b) : super(a, b, true) {}
+
+    virtual ~CosineDistances() {}
```
Comment (averbukh) [inline cpp/daal/src/algorithms/service_kernel_math.h:319]: Conflicts with public inheritance (override)

File: cpp/oneapi/dal/algo/knn/backend/cpu/infer_kernel_brute_force.cpp [hunk 10]
```diff
@@ -0,0 +1,148 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/knn/backend/cpu/infer_kernel_brute_force.cpp:2]: 2021

File: cpp/oneapi/dal/algo/knn/backend/cpu/infer_kernel_kd_tree.cpp [hunk 11]
```diff
@@ -0,0 +1,136 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/knn/backend/cpu/infer_kernel_kd_tree.cpp:2]: 2021

File: cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_brute_force_search_dpc.cpp [hunk 12]
```diff
@@ -0,0 +1,54 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_brute_force_search_dpc.cpp:2]: 2021

File: cpp/oneapi/dal/algo/knn/backend/gpu/train_kernel_brute_force_search_dpc.cpp [hunk 13]
```diff
@@ -0,0 +1,53 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/knn/backend/gpu/train_kernel_brute_force_search_dpc.cpp:2]: 2021

File: examples/oneapi/cpp/source/knn/knn_search_brute_force_dense_batch.cpp [hunk 14]
```diff
@@ -0,0 +1,47 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline examples/oneapi/cpp/source/knn/knn_search_brute_force_dense_batch.cpp:2]: 2021

File: examples/oneapi/cpp/source/knn/knn_search_brute_force_dense_batch.cpp [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+namespace knn = dal::knn;
+
+int main(int argc, char const *argv[]) {
+    const auto train_data_file_name = get_data_path("k_nearest_neighbors_train_data.csv");
+    const auto query_data_file_name = get_data_path("k_nearest_neighbors_test_data.csv");
+
+    const auto x_train = dal::read<dal::table>(dal::csv::data_source{ train_data_file_name });
+    const auto x_query = dal::read<dal::table>(dal::csv::data_source{ query_data_file_name });
+
+    using cosine_desc_t = typename dal::cosine_distance::descriptor<float>;
```
Comment (averbukh) [inline examples/oneapi/cpp/source/knn/knn_search_brute_force_dense_batch.cpp:32]: Why you need typename here?

File: examples/oneapi/cpp/source/knn/knn_search_brute_force_dense_batch.cpp [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    const auto cosine_desc = cosine_desc_t{};
+
+    const std::size_t neighbors_count = 6;
+    const auto knn_desc =
+        knn::descriptor<float, knn::method::brute_force, knn::task::search, cosine_desc_t>(
```
Comment (averbukh) [inline examples/oneapi/cpp/source/knn/knn_search_brute_force_dense_batch.cpp:37]: Does deduction work for cosine_desc_t?

**Comment by PetrovKP** [comment]
http://intel-ci.intel.com/ebcf0caf-ac64-f1a4-b37c-8cdcd4b723a1

File: cpp/oneapi/dal/algo/cosine_distance/common.hpp [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+using v1::is_valid_task_v;
+
+} // namespace detail
+
+namespace v1 {
+
+/// @tparam Float  The floating-point type that the algorithm uses for
+///                intermediate computations. Can be :expr:`float` or
+///                :expr:`double`.
+/// @tparam Method Tag-type that specifies an implementation of algorithm. Can
```
Comment (outoftardis) [inline cpp/oneapi/dal/algo/cosine_distance/common.hpp:100]:
```suggestion
/// @tparam Method Tag-type that specifies the implementation of the algorithm. Can
```

File: examples/oneapi/cpp/source/knn/knn_search_brute_force_dense_batch.cpp (hunk 16 above)
Comment (PetrovKP) [reply examples/oneapi/cpp/source/knn/knn_search_brute_force_dense_batch.cpp:37]: when I specialize other template parameters: float, knn::method::brute_force, knn::task::search - doesn't work. when the default template parameters are used, everything works.

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i (hunk 5 above)
Comment (PetrovKP) [reply cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i:85]: How was it better before?

**Review by KulikovNikita — APPROVED** [review]

**Pull request merged**